### ISSUE TITLE
`-x` space_suffix not working anymore in latest version.  

### ISSUE THREAD

**Issue by daenny** [catkin/catkin_tools#545]
[//]: # (Thank you for reporting an issue for catkin_tools!)

[//]: # (Please fill out the details below so that your issue can be resolved quickly!)
[//]: # (If you want to suggest a feature or start a discussion, feel free to delete them.)

Setting a `space_suffix` is broken in the latest version. When adding a `space_suffix` to the the catkin profile, nothing is changed in the configuration.   

### System Info

* Operating System: `Linux f717aae6a030 4.15.0-47-generic #50~16.04.1-Ubuntu SMP Fri Mar 15 16:06:21 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux` (ros-kinetic-base docker-image)
* Python Version: `Ptython 2.7.12`
* Version of catkin_tools: `catkin --version` OR *Git revision*
```
catkin_tools 0.4.5 (C) 2014-2019 Open Source Robotics Foundation
catkin_tools is released under the Apache License, Version 2.0 (http://www.apache.org/licenses/LICENSE-2.0)
```
* ROS Distro: `echo $ROS_DISTRO` melodic

### Build / Run Issue

[//]: # (If you are migrating from catkin_make, please follow the migration guide:)
[//]: # (http://catkin-tools.readthedocs.org/en/latest/migration.html)

[//]: # (Please also check for solved issues here:)
[//]: # (http://catkin-tools.readthedocs.org/en/latest/troubleshooting.html)

[//]: # (And check for open issues here:)
[//]: # (https://github.com/catkin/catkin_tools/issues?q=is%3Aopen+is%3Aissue+label%3Abug)

[//]: # (If you have an issue building and running code, please verify the following:)

* [x] Not applicable
* [ ] Works with `catkin_make`
* [ ] Works with `catkin_make_isolated --merge`
* [ ] Works with `catkin build`
* [ ] Works with `catkin build -p1`
* [ ] I did  not `read this`

### Expected Behavior

[//]: # (Please describe what you expected or think should happen.)
Adding a space_suffix should add the suffix to all standard spaces. 
```
catkin config -x _debug
-----------------------------------------------------------------
Profile:                     default
Extending:             [env] /opt/ros/kinetic
Workspace:                   /root/ros/kinetic/test
-----------------------------------------------------------------
Build Space:       [missing] /root/ros/kinetic/test/build_debug
Devel Space:       [missing] /root/ros/kinetic/test/devel_debug
Install Space:      [unused] /root/ros/kinetic/test/install_debug
Log Space:         [missing] /root/ros/kinetic/test/logs_debug
Source Space:       [exists] /root/ros/kinetic/test/src
DESTDIR:            [unused] None
-----------------------------------------------------------------
Devel Space Layout:          linked
Install Space Layout:        None
-----------------------------------------------------------------
Additional CMake Args:       -DCMAKE_EXPORT_COMPILE_COMMANDS=1
Additional Make Args:        None
Additional catkin Make Args: None
Internal Make Job Server:    True
Cache Job Environments:      False
-----------------------------------------------------------------
Whitelisted Packages:        None
Blacklisted Packages:        None
-----------------------------------------------------------------
Workspace configuration appears valid.
```

### Actual Behavior
```bash
$ catkin config
-----------------------------------------------------------------
Profile:                     default
Extending:             [env] /opt/ros/kinetic
Workspace:                   /root/ros/kinetic/test
-----------------------------------------------------------------
Build Space:       [missing] /root/ros/kinetic/test/build
Devel Space:       [missing] /root/ros/kinetic/test/devel
Install Space:      [unused] /root/ros/kinetic/test/install
Log Space:         [missing] /root/ros/kinetic/test/logs
Source Space:       [exists] /root/ros/kinetic/test/src
DESTDIR:            [unused] None
-----------------------------------------------------------------
Devel Space Layout:          linked
Install Space Layout:        None
-----------------------------------------------------------------
Additional CMake Args:       -DCMAKE_EXPORT_COMPILE_COMMANDS=1
Additional Make Args:        None
Additional catkin Make Args: None
Internal Make Job Server:    True
Cache Job Environments:      False
-----------------------------------------------------------------
Whitelisted Packages:        None
Blacklisted Packages:        None
-----------------------------------------------------------------
Workspace configuration appears valid.
-----------------------------------------------------------------
```
Setting a space_suffix 
```
catkin config -x _debug
-----------------------------------------------------------------
Profile:                     default
Extending:             [env] /opt/ros/kinetic
Workspace:                   /root/ros/kinetic/test
-----------------------------------------------------------------
Build Space:       [missing] /root/ros/kinetic/test/build
Devel Space:       [missing] /root/ros/kinetic/test/devel
Install Space:      [unused] /root/ros/kinetic/test/install
Log Space:         [missing] /root/ros/kinetic/test/logs
Source Space:       [exists] /root/ros/kinetic/test/src
DESTDIR:            [unused] None
-----------------------------------------------------------------
Devel Space Layout:          linked
Install Space Layout:        None
-----------------------------------------------------------------
Additional CMake Args:       -DCMAKE_EXPORT_COMPILE_COMMANDS=1
Additional Make Args:        None
Additional catkin Make Args: None
Internal Make Job Server:    True
Cache Job Environments:      False
-----------------------------------------------------------------
Whitelisted Packages:        None
Blacklisted Packages:        None
-----------------------------------------------------------------
Workspace configuration appears valid.
```

[//]: # (Please describe what actually happens.)

[//]: # (Please put long CLI output into Gists: http://gist.github.com)
[//]: # (Use `VERBOSE=1 catkin build -p1 -v` for verbose output)
[//]: # (Please post `logs` directory if available.)

### Steps to Reproduce the Issue

[//]: # (Please give a series of command-line commands which cause the problem.)
[//]: # (Include a checkout URI for the workspace so that we can reproduce the issue.)
[//]: # (If we can't reproduce it on our own machines, we can't fix it!)

```bash
see above
```

**Comment by racko** [comment]
Cannot reproduce:
```
$ pip2 install --upgrade --user catkin_tools
...
Successfully installed catkin-pkg-0.4.11 catkin-tools-0.4.5 osrf-pycommon-0.1.6 trollius-2.2
$ ~/.local/bin/catkin --version
catkin_tools 0.4.5 (C) 2014-2019 Open Source Robotics Foundation
catkin_tools is released under the Apache License, Version 2.0 (http://www.apache.org/licenses/LICENSE-2.0)
---
Using Python 2.7.16 (default, Mar 11 2019, 18:59:25) [GCC 8.2.1 20181127]
$ cd /tmp
$ mkdir ws
$ cd ws
$ ~/.local/bin/catkin init       
Initializing catkin workspace in `/tmp/ws`.
--------------------------------------------
Profile:                     default
Extending:                   None
Workspace:                   /tmp/ws
--------------------------------------------
Build Space:       [missing] /tmp/ws/build
Devel Space:       [missing] /tmp/ws/devel
Install Space:      [unused] /tmp/ws/install
Log Space:         [missing] /tmp/ws/logs
Source Space:      [missing] /tmp/ws/src
DESTDIR:            [unused] None
--------------------------------------------
Devel Space Layout:          linked
Install Space Layout:        None
--------------------------------------------
Additional CMake Args:       None
Additional Make Args:        None
Additional catkin Make Args: None
Internal Make Job Server:    True
Cache Job Environments:      False
--------------------------------------------
Whitelisted Packages:        None
Blacklisted Packages:        None
--------------------------------------------


--------------------------------------------
WARNING: Your workspace is not extending
any other result space, but it is set to
use a `linked` devel space layout. This
requires the `catkin` CMake package in your
source space in order to be built.
--------------------------------------------
WARNING: Source space `/tmp/ws/src` does
not yet exist.
--------------------------------------------

racko@yodada /tmp/ws % ~/.local/bin/catkin config -x _debug
--------------------------------------------------
Profile:                     default
Extending:                   None
Workspace:                   /tmp/ws
--------------------------------------------------
Build Space:       [missing] /tmp/ws/build_debug
Devel Space:       [missing] /tmp/ws/devel_debug
Install Space:      [unused] /tmp/ws/install_debug
Log Space:         [missing] /tmp/ws/logs_debug
Source Space:      [missing] /tmp/ws/src
DESTDIR:            [unused] None
--------------------------------------------------
Devel Space Layout:          linked
Install Space Layout:        None
--------------------------------------------------
Additional CMake Args:       None
Additional Make Args:        None
Additional catkin Make Args: None
Internal Make Job Server:    True
Cache Job Environments:      False
--------------------------------------------------
Whitelisted Packages:        None
Blacklisted Packages:        None
--------------------------------------------------


--------------------------------------------------
WARNING: Your workspace is not extending any
other result space, but it is set to use a
`linked` devel space layout. This requires the
`catkin` CMake package in your source space in
order to be built.
--------------------------------------------------
WARNING: Source space `/tmp/ws/src` does not yet
exist.
--------------------------------------------------
```

Changing the prefix after that does not work:
```
$ ~/.local/bin/catkin config -x _foo  
--------------------------------------------------
Profile:                     default
Extending:                   None
Workspace:                   /tmp/ws
--------------------------------------------------
Build Space:       [missing] /tmp/ws/build_debug
Devel Space:       [missing] /tmp/ws/devel_debug
Install Space:      [unused] /tmp/ws/install_debug
Log Space:         [missing] /tmp/ws/logs_debug
Source Space:      [missing] /tmp/ws/src
DESTDIR:            [unused] None
--------------------------------------------------
Devel Space Layout:          linked
Install Space Layout:        None
--------------------------------------------------
Additional CMake Args:       None
Additional Make Args:        None
Additional catkin Make Args: None
Internal Make Job Server:    True
Cache Job Environments:      False
--------------------------------------------------
Whitelisted Packages:        None
Blacklisted Packages:        None
--------------------------------------------------


--------------------------------------------------
WARNING: Your workspace is not extending any
other result space, but it is set to use a
`linked` devel space layout. This requires the
`catkin` CMake package in your source space in
order to be built.
--------------------------------------------------
WARNING: Source space `/tmp/ws/src` does not yet
exist.
--------------------------------------------------

```
The help says
>  -x SPACE_SUFFIX, --space-suffix SPACE_SUFFIX
>                        Suffix for build, devel, and install space if they are
>                        not otherwise explicitly set.

As far as I can tell, `-x` actually *sets* the spaces and does not set them if they are already set.

**Comment by daenny** [comment]
Ok I think the problem might be that I did a `catkin profile add test --copy-from default`. But this used to work before.

**Comment by Matthew-Reynolds** [comment]
@racko I believe the case you present is actually one of very few cases in which `--space-suffix` works.

Whenever a context is saved, which happens whenever `catkin profile add`, `catkin profile rename`, `catkin build --save-config`, or any `catkin config` commands are run, the default space values are written to the profile's `config.yaml`. Since a value for each space now exists, the `--space-suffix` no longer affects them. The issue is that these values are default values, not "explicitly set" values, and therefore should still be eligible for `--space-suffix` to affect them.

In short, the issue is that the tool should treat default values and no values the same way, but currently does not.

I've proposed a fix for this behaviour in #560

**Comment by Matthew-Reynolds** [comment]
@racko in your case, since you did not run a `catkin config` to read back the config summary before attempting to set the space suffix, you never saved the default values to the `config.yaml`.

Even commands such as `catkin config` and `catkin profile add` write the default values to the file and break the `--space-suffix` argument, meaning it's impossible to use `--space-suffix` in a new profile or on a workspace that isn't essentially brand new.

**Comment by mikepurvis** [comment]
Closed via #560.
